VAO-WIP (#83)
org-arl/unet-blog: _posts/2025-05-11-running-hardware-in-the-loop-simulations-using-subnero-modems.md

## Typical Use Cases
- **Researchers** testing new communication algorithms (e.g. physical layer, network layer etc.) with realistic conditions
- **Commercial users** simulating data harvesting from sensors or AUVs in unknown environments
- **Defense teams** validating mission scenarios where real-world testing is restricted or classified

## Step-by-Step Guide to HIL Simulation

Before beginning the simulation, ensure you have the following setup:
- Two Subnero modems running UnetStack v5
- A laptop or mini-PC with Julia installed, capable of running the Virtual Acoustic Ocean (VAO) simulator
- An Ethernet switch to connect all devices to the same network

All three components—both modems and the VAO-running machine—should be on the same subnet and able to reach each other via IP

> NOTE: It is strongly recommended to use an isolated network switch (without internet access) to avoid interference from unrelated network traffic during simulation.

Here is a reference block diagram of a typical HIL simulation setup:

TODO:

1. Install VAO using Julia's package manager. In a Julia session, run:

```
using Pkg

> Pkg.add("VirtualAcousticOcean")

Author: ```suggestion
Pkg.add("UnderwaterAcoustics")
Pkg.add("VirtualAcousticOcean")
```

This is where HIL simulations shine. By keeping your real hardware in the loop, you:
- Test the actual modem hardware, drivers, and firmware
- Verify timing behavior and interface compatibility
- Run real world network software with simulated acoustic links

If your modem is software-defined and open-architecture, such as Subnero's modems, HIL offers a development experience that closely mirrors actual deployment.

## The Foundation: Subnero Modems + Virtual Acoustic Ocean

Subnero's modems run UnetStack, an open-architecture, software-defined acoustic modem (SDOAM) framework. This allows users to:
- Customize physical and network layer behaviors
- Deploy the exact same software in simulation and in the ocean
- Extend or override components with user-defined protocols

[Virtual Acoustic Ocean (VAO)](https://github.com/org-arl/VirtualAcousticOcean.jl) is an open-source acoustic channel simulator developed in Julia. It allows users to:

> - Simulate channel propagation using models like Bellhop, Kraken, Pekeris RAT

Author: ```suggestion
- Simulate channel propagation using standard models
```

- Set up nodes with position, protocol stack, and acoustic properties

> - Run interactive or batch-mode simulations over UDP/IP

Author: ```suggestion
- Run interactive or batch-mode simulations
```

## Typical Use Cases
- **Researchers** testing new communication algorithms (e.g. physical layer, network layer etc.) with realistic conditions
- **Commercial users** simulating data harvesting from sensors or AUVs in unknown environments
- **Defense teams** validating mission scenarios where real-world testing is restricted or classified

## Step-by-Step Guide to HIL Simulation

Before beginning the simulation, ensure you have the following setup:
- Two Subnero modems running UnetStack v5
- A laptop or mini-PC with Julia installed, capable of running the Virtual Acoustic Ocean (VAO) simulator
- An Ethernet switch to connect all devices to the same network

All three components—both modems and the VAO-running machine—should be on the same subnet and able to reach each other via IP

> NOTE: It is strongly recommended to use an isolated network switch (without internet access) to avoid interference from unrelated network traffic during simulation.

Here is a reference block diagram of a typical HIL simulation setup:

TODO:

1. Install VAO

```

> git clone https://github.com/org-arl/VirtualAcousticOcean.jl.git

Author: ```suggestion
using Pkg
Pkg.add("VirtualAcousticOcean")
```

cd VirtualAcousticOcean.jl
julia --project=.
```

> Inside Julia's package manager:

Author: ```suggestion
```

> 

Author: ```suggestion
```

```

> pkg> instantiate

Author: ```suggestion
```

> ```

Author: ```suggestion
```

## Why Hardware-in-the-Loop (HIL)?

While pure software simulation (where all nodes run on a single computer) is useful, it falls short in certain cases, particularly when real-time behavior, hardware driver integration, and system compatibility are important.

Imagine you are a researcher developing a large scale underwater network. You want to validate not only your algorithms but also the performance of the modems you plan to use in the field. Before investing time and money into an offshore campaign, you want confidence that your system behaves correctly, both in logic and in timing. You also want to try out multiple configurations quickly.

This is where HIL simulations shine. By keeping your real hardware in the loop, you:

- Test the actual modem hardware, drivers, and firmware.
- Verify timing behavior and interface compatibility of your algorithms in the modems.
- Run real world network software with simulated acoustic links.

If your modem is software-defined and open-architecture, such as Subnero's modems, HIL offers a development experience that closely mirrors actual deployment.

> ## The Foundation: Subnero Modems + Virtual Acoustic Ocean

Author: ```suggestion
## The Foundation: UnetStack v5 + Virtual Acoustic Ocean
```

> Subnero’s modems run UnetStack, a fully software-defined, open-architecture modem framework. This enables users to customize physical and network layers, extend built-in components, and use the exact same software stack in both simulated and real-world deployments.

Author: ```suggestion
Modems running UnetStack v5 — such as those offered by Subnero — are fully software-defined and follow an open architecture. This enables users to customize physical and network layers, extend built-in components, and use the exact same software stack in both simulated and real-world deployments.
```

```toml
[input]
analoginterface = "UASP2DAQ"
ip = "192.168.42.10"
port = 9809
```

> Ensure this matches the VAO machine's IP and each modem uses the correct port (e.g. 9809).

**Verify the setup via modem shell:**

```shell
bb
```

**Expected output:**

```
> bb
« Baseband »

Baseband service for UASP2DAQ.AnalogInterface
...
```

Also check modem logs for an entry similar to:

```
1750584190212	INFO	UASP2DAQ@27:UASP2DAQ@1:57	Connecting to UASP2 DAQ at 192.168.42.10:9809...
```

That's it. Now you have 2 nodes deployed 1 km apart and connected through VAO. With the nodes now connected via VAO and properly configured, you can begin interactive testing and messaging between the modems.

---

### 4. Set Transmit Power

Since the modems are deployed 1 km apart, set full transmit power:

```shell
plvl 0
```

---

### 5. Test the Link

Send a test message from either of the modem's shell:

```shell
tell 0 "Hello Sea"
```

You should see the message on the receiving modem’s shell, confirming that the simulated link is working.

---

## UnetCube (UnetStack OEM Edition) + VAO

> If you do not have Subnero modems, you can still simulate the system using UnetCube (UnetStack OEM Edition) on single-board computers (e.g., Jetson Orin Nano). The simulation behavior will be the same, enabling rapid prototyping.

Author: ```suggestion
If you do not have Subnero modems, you can still simulate the system using UnetCube (UnetStack OEM Edition) on single-board computers (e.g., Jetson Orin Nano). The simulation behavior will be the same, enabling rapid prototyping.

<figure> <img src="../assets/img/unetcube.jpg"></figure>

```

## Step-by-Step Guide to HIL Simulation

Before beginning the simulation, ensure you have the following setup:

- Two Subnero M25M series modems running UnetStack v5.
- A laptop or mini-PC with Julia installed, capable of running the Virtual Acoustic Ocean (VAO) simulator.
- An Ethernet switch to connect all devices to the same network.

All components — modems and the VAO-running machine — should be on the same subnet to be able to reach each other using their IP address.

> **TIP:** Use an isolated network switch (without internet access) to avoid interference from unrelated network traffic during simulation.

> Note: This example uses Subnero M25M (MF-band) modems, but the same steps apply to Subnero's LF and HF modems. The simulation script provided is just one example — please refer to the VAO GitHub repository for the latest examples, which may differ. This guide shows how to connect two Subnero modems, but more nodes can be added by extending the script with additional addnode! lines.

---

### 1. Install VAO and Packages

At a Julia terminal:

```julia
using Pkg
Pkg.add("VirtualAcousticOcean")
Pkg.add("UnderwaterAcoustics")

> Pkg.add("Sockets")

Reviewer: We might want to add version specifiers here:

```julia
Pkg.add(name = "VirtualAcousticOcean", version = "0.6")
...

```

@veronicaSubnero had a brief moment of confusion on Friday because she had old versions of these packages installed, and then the code snippets below didn't work. 

Alternatively, we could complement `2-node-network.jl` with a `Project.toml` file and ask users to download both.

Author: Added the version for now.

```

---

### 2. Add Example Scenario

> Copy the `2-node-network.jl` file from the [GitHub repo](https://github.com/org-arl/VirtualAcousticOcean.jl) examples folder to your working directory.

Reviewer 2: Users may be confused because the `2-node-network.jl` shown in the example content is different from the `2-node-network.jl` in the mentioned directory. If they run the version from the directory: (1) it won’t be clear to users whether VAO is running (2) it's not using sockets, and it would not work.

**Example content:**

```julia
using VirtualAcousticOcean
using UnderwaterAcoustics
using Sockets

env = UnderwaterEnvironment(seabed=SandyClay, bathymetry=40.0)
pm = PekerisRayTracer(env)
sim = Simulation(pm, 24000.0)
addnode!(sim, (0.0, 0.0, -10.0), UASP2, 9809, ip"0.0.0.0")
addnode!(sim, (1000.0, 0.0, -10.0), UASP2, 9819, ip"0.0.0.0")
run(sim)

println("Simulation running with these nodes:")
for (idx, node) in enumerate(sim.nodes)
    println("  - Node $(idx) at position $(node.pos) receiving on UDP port $(node.conn.port)")
end

wait()
```

**Run it from the terminal:**

```bash
> julia 2-node-network.jl
Simulation running with these nodes:
  - Node 1 at position (0.0, 0.0, -10.0) receiving on UDP port 9809
  - Node 2 at position (1000.0, 0.0, -10.0) receiving on UDP port 9819
```

---

### 3. Configure Subnero Modems

Each modem needs to be configured to redirect its baseband interface to VAO. Access each modem’s web UI and upload this `modem.toml` to the Scripts folder:

```toml

> [input]
> analoginterface = "UASP2DAQ"
> ip = "192.168.42.10"
> port = 9809

Reviewer 2: ```suggestion
[input]
analoginterface = "UASP2DAQ"
ip = "192.168.42.10"
port = 9809

[output]
analoginterface = "UASP2DAQ"
```

When testing without [output], the modem is not be connected to VAO and is likely to be running the [output] modem's configuration. The following error can be seen:

```
MethodError(BB.AnalogInterfaces.setinputgaincontroller, (UASP2DAQ.AnalogInterface(), PillioDAQ.AnalogInterface()), 0x00000000000068e6)
--
  |   |   | Stack trace:
  |   |   | BB.CombinedDAQ.AnalogInterface(inp::BB.UASP2DAQ.AnalogInterface, out::BB.PillioDAQ.AnalogInterface) at CombinedDAQ.jl:11
  |   |   | select_analoginterface(fs::Float64) at initrc.jl:221
  |   |   | modemsetup(c::Fjage.SlaveContainer{Fjage.RealTimePlatform}) at initrc.jl:90
  |   |   | run(args::Vector{String}; blocking::Bool) at initrc.jl:50
  |   |   | run at initrc.jl:17 [inlined]
 ```
 
 Hence, we should define the [output]

```

> Ensure this matches the VAO machine's IP and each modem uses the correct port (e.g. 9809).

**Verify the setup via modem shell:**

```shell
bb
```

**Expected output:**

```
> bb
« Baseband »

Baseband service for UASP2DAQ.AnalogInterface
...
```

Also check modem logs for an entry similar to:

```
1750584190212	INFO	UASP2DAQ@27:UASP2DAQ@1:57	Connecting to UASP2 DAQ at 192.168.42.10:9809...
```

That's it. Now you have 2 nodes deployed 1 km apart and connected through VAO. With the nodes now connected via VAO and properly configured, you can begin interactive testing and messaging between the modems.

---

### 4. Set Transmit Power

Since the modems are deployed 1 km apart, set full transmit power:

```shell
plvl 0
```

---

### 5. Test the Link

Send a test message from either of the modem's shell:

```shell

> tell 0 "Hello Sea"

Reviewer 2: ```suggestion
tell 0, "Hello Sea"
```

> tell 0 "Hello Sea"

Author: ```suggestion
tell 0, "Hello Sea"
```

---

### 1. Install VAO and Packages

At a Julia terminal:

```julia
using Pkg
Pkg.add("VirtualAcousticOcean")
Pkg.add("UnderwaterAcoustics")
Pkg.add("Sockets")
```

---

### 2. Add Example Scenario

Copy the `2-node-network.jl` file from the [GitHub repo](https://github.com/org-arl/VirtualAcousticOcean.jl) examples folder to your working directory.

**Example content:**

```julia
using VirtualAcousticOcean
using UnderwaterAcoustics
using Sockets

env = UnderwaterEnvironment(seabed=SandyClay, bathymetry=40.0)
pm = PekerisRayTracer(env)
sim = Simulation(pm, 24000.0)
addnode!(sim, (0.0, 0.0, -10.0), UASP2, 9809, ip"0.0.0.0")
addnode!(sim, (1000.0, 0.0, -10.0), UASP2, 9819, ip"0.0.0.0")
run(sim)

println("Simulation running with these nodes:")
for (idx, node) in enumerate(sim.nodes)
    println("  - Node $(idx) at position $(node.pos) receiving on UDP port $(node.conn.port)")
end

wait()
```

> 

Author: ```suggestion

> Note: The example shown here is representative. The version available in the [VAO GitHub repository](https://github.com/org-arl/VirtualAcousticOcean.jl) may differ slightly depending on updates or version changes.
```

env = UnderwaterEnvironment(seabed=SandyClay, bathymetry=40.0)
pm = PekerisRayTracer(env)
sim = Simulation(pm, 24000.0)
addnode!(sim, (0.0, 0.0, -10.0), UASP2, 9809, ip"0.0.0.0")
addnode!(sim, (1000.0, 0.0, -10.0), UASP2, 9819, ip"0.0.0.0")
run(sim)

println("Simulation running with these nodes:")
for (idx, node) in enumerate(sim.nodes)
    println("  - Node $(idx) at position $(node.pos) receiving on UDP port $(node.conn.port)")
end

wait()
```

> Note: The example shown here is representative. The version available in the [VAO GitHub repository](https://github.com/org-arl/VirtualAcousticOcean.jl) may differ slightly depending on updates or version changes.
**Run it from the terminal:**

```bash
> julia 2-node-network.jl
Simulation running with these nodes:
  - Node 1 at position (0.0, 0.0, -10.0) receiving on UDP port 9809
  - Node 2 at position (1000.0, 0.0, -10.0) receiving on UDP port 9819
```

---

### 3. Configure Subnero Modems

Each modem needs to be configured to redirect its baseband interface to VAO. Access each modem’s web UI and upload this `modem.toml` to the Scripts folder:

```toml
[input]
analoginterface = "UASP2DAQ"
ip = "192.168.42.10"
port = 9809
```

> > Ensure this matches the VAO machine's IP and each modem uses the correct port (e.g. 9809).

Author: ```suggestion
> Ensure this matches the VAO machine's IP and each modem uses the correct port (e.g. 9809). Additionally, the `[output]` section is optional and typically used when configuring a physical modem.
```
